File: solutions/haskell/perfect-numbers/1/src/PerfectNumbers.hs
module PerfectNumbers (classify, Classification(..)) where

data Classification = Deficient | Perfect | Abundant deriving (Eq, Show)

aliquotSum :: Int -> [Int]
aliquotSum n = filter (\m -> n `mod` m == 0) [1 .. n-1]

classify :: Int -> Maybe Classification
classify n
 | n <= 0       = Nothing
 | n == aliquit = Just Perfect
 | aliquit < n  = Just Deficient
 | aliquit > n  = Just Abundant
 where
   aliquit = sum (aliquotSum n)

File: solutions/erlang/perfect-numbers/1/src/perfect_numbers.erl
-module(perfect_numbers).
-export([classify/1]).
-import(lists, [seq/2, sum/1]).

classify(N) when N =< 0 -> erlang:error("Input is not positive integer");
classify(N) ->
  Divisors = [X || X <- lists:seq(1, N div 2), N rem X == 0],
  nicomachusClass(sum(Divisors), N).

nicomachusClass(AliquotSum, N) when AliquotSum > N -> abundant;
nicomachusClass(AliquotSum, N) when AliquotSum < N -> deficient;
nicomachusClass(AliquotSum, N) when AliquotSum == N -> perfect.
